packages/pubsub/src/subscription.js

```diff
@@ -592,157 +442,398 @@ Subscription.prototype.decorateMessage_ = function(message) {
  *   var apiResponse = data[0];
  * });
  */
-Subscription.prototype.delete = function(callback) {
+Subscription.prototype.delete = function(gaxOpts, callback) {
   var self = this;
 
-  callback = callback || common.util.noop;
-
-  var protoOpts = {
-    service: 'Subscriber',
-    method: 'deleteSubscription'
-  };
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
 
   var reqOpts = {
     subscription: this.name
   };
 
-  this.parent.request(protoOpts, reqOpts, function(err, resp) {
-    if (err) {
-      callback(err, resp);
-      return;
+  this.request({
+    client: 'subscriberClient',
+    method: 'deleteSubscription',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, function(err, resp) {
+    if (!err) {
+      self.removeAllListeners('message');
+      self.close();
     }
 
-    self.closed = true;
-    self.removeAllListeners();
-
-    callback(null, resp);
+    callback(err, resp);
   });
 };
 
 /**
- * Pull messages from the subscribed topic. If messages were found, your
- * callback is executed with an array of message objects.
+ * Flushes internal queues. These can build up if a user attempts to ack/nack
+ * while there is no connection pool (e.g. after they called close).
  *
- * Note that messages are pulled automatically once you register your first
- * event listener to the subscription, thus the call to `pull` is handled for
- * you. If you don't want to start pulling, simply don't register a
- * `subscription.on('message', function() {})` event handler.
+ * Typically this will only be called either after a timeout or when a
+ * connection is re-opened.
  *
- * @todo Should not be racing with other pull.
+ * Any errors that occur will be emitted via `error` events.
  *
- * @resource [Subscriptions: pull API Documentation]{@link https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull}
+ * @private
+ */
+Subscription.prototype.flushQueues_ = function() {
+  var self = this;
+
+  var acks = this.inventory_.ack;
+  var nacks = this.inventory_.nack;
+
+  if (!acks.length && !nacks.length) {
+    return;
+  }
+
+  if (this.connectionPool) {
+    this.connectionPool.acquire(function(err, connection) {
+      if (err) {
+        self.emit('error', err);
+        return;
+      }
+
+      var reqOpts = {};
+
+      if (acks.length) {
+        reqOpts.ackIds = acks;
+      }
+
+      if (nacks.length) {
+        reqOpts.modifyDeadlineAckIds = nacks;
+        reqOpts.modifyDeadlineSeconds = Array(nacks.length).fill(0);
+      }
+
+      connection.write(reqOpts);
+
+      self.inventory_.ack = [];
```

Comment:
I'm not sure there is a reliable way to do that. Per the [official docs](https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback) the optional callback you can provide to write might execute without an error despite it failing. We could listen for an error event, but there's no guarantee that we won't end up intercepting an error for a different write call.